[sw/crypto] Add second RSA padding check as a FI countermeasure

h-filali opened the pull request:
This PR adds a second RSA padding check as an FI countermeasure.

Related to #26940

jadephilipoom reviewed (changes requested):
I think this check is not quite what was meant by the TODO. In the previous code, `decode_failure` is the `or` of several values which are all 0s in the "good" case and all 1s in the "bad" case. In short:
```c
decode_failure = 0 | padding_failure | message_start_not_found | lhash_mismatch | leading_byte_nonzero;
if (decode_failure != 0) {
  return OTCRYPTO_BAD_ARGS;
}
HARDENED_CHECK_EQ(decode_failure, 0);
```
The new code does:
```c
decode_failure = 0 | padding_failure | message_start_not_found | lhash_mismatch | leading_byte_nonzero;
decode_failure |= padding_failure;
if (decode_failure != 0) {
  return OTCRYPTO_BAD_ARGS;
}
HARDENED_CHECK_EQ(decode_failure, 0);
```
Even though one of those values is called `padding_failure`, all of these checks are in fact padding checks, and the TODO meant all of them. For FI resistance the `padding failure` value is not more important than the others. More importantly, the main pressure point for FI would be the if statement, or more precisely zeroing the `decode_failure` value right before the if statement. A more effective FI-resistant check would be:
```c
decode_failure = 0 | padding_failure | message_start_not_found | lhash_mismatch | leading_byte_nonzero;
if (decode_failure != 0) {
  return OTCRYPTO_BAD_ARGS;
}
HARDENED_CHECK_EQ(decode_failure, 0);

// Repeat the padding checks for FI resistance.
decode_failure = 0 | padding_failure | message_start_not_found | lhash_mismatch | leading_byte_nonzero;
if (decode_failure != 0) {
  return OTCRYPTO_BAD_ARGS;
}
HARDENED_CHECK_EQ(decode_failure, 0);
```
It would also probably make sense to ensure that the message start index matches both times it's recovered. This is all a bit heavyweight as an FI countermeasure, but since RSA decryption is so slow anyway it almost certainly doesn't change the big picture performance-wise.

h-filali commented:
@jadephilipoom quick question, why do we have an if condition and a hardened check. Shouldn't the hardened check suffice?

`if (decode_failure != 0) {
  return OTCRYPTO_BAD_ARGS;
}
HARDENED_CHECK_EQ(decode_failure, 0);`

jadephilipoom commented:
> @jadephilipoom thanks for your review! Quick question, why do we have an if condition and a hardened check. Shouldn't the hardened check suffice?

The hardened check doesn't neatly return an error, it hits a sequence of `unimp` instructions if it fails and the program crashes. We don't want to do that to an innocent caller who made a small mistake :slightly_smiling_face:

h-filali commented on sw/device/lib/crypto/impl/rsa/rsa_padding.c:
```diff
@@ -386,6 +386,54 @@ static status_t pss_construct_h(const otcrypto_hash_digest_t message_digest,
       h_buffer);
 }
 
+static status_t rsa_padding_oaep_check(
```
Okay didn't know that. Thanks a lot!

jadephilipoom commented on sw/device/lib/crypto/impl/rsa/rsa_padding.c:
```diff
@@ -702,54 +769,22 @@ status_t rsa_padding_oaep_decode(const otcrypto_hash_mode_t hash_mode,
+      db, lhash_data, digest_wordlen, digest_bytelen, encoded_message_bytes,
+      db_bytelen, &message_start_idx1));
+
+  HARDENED_CHECK_EQ(message_start_idx0, message_start_idx1);
```
I'd recommend checking the disassembly to make sure the compiler didn't observe that these values are logically the same and "helpfully" eliminate the check for you. You can find the `.dis` file by running this query with bazel (any test target that depends on OAEP padding will work):
```
./bazelisk.sh aquery 'outputs(".*.dis", //sw/device/tests/crypto:rsa_2048_encryption_functest_fpga_cw310_rom_with_fake_keys)' | grep 'Outputs'
```
For example, if I do this on the current version and look for that hardened `if (launder32(decode_failure) != 0)` statement, I find:
```asm
launder32():
/proc/self/cwd/./sw/device/lib/base/hardened.h:251
20002c6a:       |  |               85aa                 mv      a1,a0
rsa_padding_oaep_decode():
/proc/self/cwd/sw/device/lib/crypto/impl/rsa/rsa_padding.c:741
20002c6c:       |  |           /-- c991                 beqz    a1,20002c80 <rsa_padding_oaep_decode+0x406>
20002c6e:       |  |           |   80010537             lui     a0,0x80010
20002c72:       |  |           |   ea350993             addi    s3,a0,-349 # 8000fea3 <_otbn_remote_app_run_rsa_modexp__checksum+0xdcd2c3ad>
20002c76:       \--|-----------|-- b54d                 j       20002b18 <rsa_padding_oaep_decode+0x29e>
20002c78:          |           |   c0001073             unimp
20002c7c:          |           |   c0001073             unimp
/proc/self/cwd/sw/device/lib/crypto/impl/rsa/rsa_padding.c:744
20002c80:          |        /--\-X c111                 beqz    a0,20002c84 <rsa_padding_oaep_decode+0x40a>
20002c82:          |        |  /-> 0000                 unimp
20002c84:          |        \--\-X fd7d                 bnez    a0,20002c82 <rsa_padding_oaep_decode+0x408>
```
This shows me several different branches that lead to `unimp` or returning an error (`20002b18` is a path that returns `OTCRYPTO_BAD_ARGS`) if the `decode_failure` value in `a0` and `a1` is nonzero. The "good" path continues on from the code I pasted, and the way to get there is:
- take the `beqz` branch against `a1` at `20002c6c`
- take the `beqz` branch against `a0` at `20002c80`
- don't take the `bnez` branch against `a0` at `20002c84`

Skipping any of those checks runs into an `unimp`. But if I remove the `launder32` from the `if` statement, I see:
```asm
20002c6a:       |  |           /-- c911                 beqz    a0,20002c7e <rsa_padding_oaep_decode+0x404>
20002c6c:       |  |           |   80010537             lui     a0,0x80010
20002c70:       |  |           |   ea350993             addi    s3,a0,-349 # 8000fea3 <_otbn_remote_app_run_rsa_modexp__checksum+0xdcd2c3ad>
20002c74:       \--|-----------|-- b555                 j       20002b18 <rsa_padding_oaep_decode+0x29e>
20002c76:          |           |   c0001073             unimp
20002c7a:          |           |   c0001073             unimp
/proc/self/cwd/sw/device/lib/crypto/impl/rsa/rsa_padding.c:744
20002c7e:          |        /--\-X 00000363             beqz    zero,20002c84 <rsa_padding_oaep_decode+0x40a>
20002c82:          |        |  /-> 0000                 unimp
20002c84:          |        \--\-X fe001fe3             bnez    zero,20002c82 <rsa_padding_oaep_decode+0x408>
```
Note that the last two branches now use `zero`, so they no longer provide any defensive redundancy. Sometimes, the secondary check might even disappear entirely. What's happening here is that `launder32` tells the compiler "the value here might change after this check, don't make assumptions about it when making optimizations" -- the documentation in `hardening.h` is a good read if you haven't read it already. Without that filter, the compiler sees "oh, if we take the first branch then it means the value is always zero, let me just use the zero register". For FI hardening, it's important to check the disassembly if you're doing anything other than a very formulaic pattern like an if statement with `launder32` and then a hardening macro checking the same condition. You might be fine here, but it's worth taking a look (and please post the disassembly here too!).

Hint: if you find the check has disappeared, you might need to insert the `launder32` around the message start index within the `oaep_check_padding_and_find_message_start` function.

h-filali replied:
I checked the disassembly and both the checks are in there as well as the final comparison of the two message start variables. I'll comment the relevant disassembly part as a comment in this PR.

jadephilipoom reviewed (commented):
Almost there, but we need to check the disassembly to make sure the compiler doesn't just eliminate the redundancy.

h-filali commented:
@jadephilipoom Thanks for your explanation. The relevant disassmebly part looks like this:

```asm
launder32():
/proc/self/cwd/./sw/device/lib/base/hardened.h:251
20002d2c:                    85aa                       mv      a1,a0
rsa_padding_oaep_check():
/proc/self/cwd/sw/device/lib/crypto/impl/rsa/rsa_padding.c:429
20002d2e:                /-- c991                       beqz    a1,20002d42 <rsa_padding_oaep_check+0x1be>
20002d30:                |   80010537           lui     a0,0x80010
20002d34:                |   ea350513           addi    a0,a0,-349 # 8000fea3 <_otbn_remote_app_run_rsa_modexp__checksum+0xdcd2c3ad>
20002d38:          /-----|-- a811                       j       20002d4c <rsa_padding_oaep_check+0x1c8>
20002d3a:          |     |   c0001073           unimp
20002d3e:          |     |   c0001073           unimp
/proc/self/cwd/sw/device/lib/crypto/impl/rsa/rsa_padding.c:432
20002d42:          |  /--\-X c111                       beqz    a0,20002d46 <rsa_padding_oaep_check+0x1c2>
20002d44:          |  |  /-> 0000                       unimp
20002d46:          |  \--\-X fd7d                       bnez    a0,20002d44 <rsa_padding_oaep_check+0x1c0>
20002d48:          |         73900513           li      a0,1849
20002d4c:          \-------> 00003297           auipc   t0,0x3
20002d50:                    86e282e7           jalr    t0,-1938(t0) # 200055ba <OUTLINED_FUNCTION_5>
/proc/self/cwd/sw/device/lib/crypto/impl/rsa/rsa_padding.c:435
20002d54:                    8082                       ret
20002d56:                    c0001073           unimp
20002d5a:                    c0001073           unimp
```

The `beqz` check on `20002d2e` checks `a1` and not the `zero` reg so we should be good here.

jadephilipoom commented:
> @jadephilipoom Thanks for your explanation. The relevant disassmebly part looks like this:

This isn't the check I commented on, it's the `decode_failure != 0` check that I used as an example. That code follows a very standard pattern so I don't need to see the disassembly; I was just using that to demonstrate how to check that redundant checks are preserved. The check I commented on is the one for equal _message indices_, which does not follow a well-established standard pattern, so it's worth checking the disassembly for that one.

h-filali commented:
Sure, this is the disassembly for the index comparison:

```asm
/proc/self/cwd/sw/device/lib/crypto/impl/rsa/rsa_padding.c:782
20002af0:       |  |  |  |         f8040693             addi    a3,s0,-128
20002af4:       |  |  |  |         5e8c                 lw      a1,56(a3)
20002af6:       |  |  |  |         5ac8                 lw      a0,52(a3)
20002af8:       |  |  |  |  /----- 00a58363             beq     a1,a0,20002afe <rsa_padding_oaep_decode+0x284>
20002afc:       |  |  |  |  |  /-> 0000                 unimp
20002afe:       |  |  |  |  \--\-X fea59fe3             bne     a1,a0,20002afc <rsa_padding_oaep_decode+0x282>
```

looks good to me.

andrea-caforio commented on sw/device/lib/crypto/impl/rsa/rsa_padding.c:
```diff
@@ -704,53 +770,23 @@ status_t rsa_padding_oaep_decode(const otcrypto_hash_mode_t hash_mode,
-  HARDENED_CHECK_EQ(decode_failure, 0);
-
-  // TODO: re-check the padding as an FI hardening measure?
+  uint32_t message_start_idx0 = 0;
```
Maybe marking the variables as `volatile` prevents some of the compiler optimizations?

h-filali replied:
Good shout, changed it to volatile. Disassembly still looks good.

h-filali commented:
Thanks @andrea-caforio for your comments. Appreciate it!

I addressed all the comments now. I think if CI passes this can probably be merged. What do you think @nasahlpa @johannheyszl
